teps/0155-whenexpressions-in-step.md

```diff
@@ -0,0 +1,110 @@
+   - [Goals](#goals)
+   - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+   - [When Expressions Evaluation](#when-expressions-evaluation)
+   - [Supported Expression Languages](#supported-expression-languages)
+- [Test Plan](#test-plan)
+- [Alternatives](#alternatives)
+
+## Summary
+This TEP proposes the addition of When Expressions to individual Steps within a Task. This will allow for fine-grained control over the execution of steps based on conditions like the status of previous step results. This enhancement will provide greater flexibility and control in defining task logic, enabling more complex and dynamic workflows.
+## Motivation
+Currently, Tekton Tasks execute steps sequentially without the ability to control their execution based on conditions within the Task itself. This limitation restricts the ability to create dynamic workflows that adapt to different scenarios or respond to the outcomes of previous steps.
+By introducing When Expressions to Steps, users will be able to:
+ - **Create conditional workflows**: Execute steps only when specific conditions are met, such as the success or failure of a previous step, the value of a result, or the presence of a specific parameter.
+ - **Improve error handling**: Implement robust error handling strategies by executing specific steps only when errors occur in previous steps.
+ - **Enhance reusability**: Develop more versatile Tasks that can adapt their behavior based on different input parameters or runtime conditions.
+
+## Goals
+ - Enable the use of When Expressions to control the execution of individual Steps within a Task.
+Support conditions based on:
+   - **Conditions for Task When Expressions**: This includes conditions supported by the existing When Expression implementation for Tasks, ensuring consistency and a familiar experience for users.
+   - **Results**: Utilize the output values of previous steps in the decision process.
+Maintain consistency with the existing When Expressions functionality used in Tasks and Pipelines.
+## Non-Goals
+ - Introduce new expression syntax or operators beyond what is already available in When Expressions.
+
+## Proposal
+We propose extending the Step definition in Tekton Tasks to include an optional when field, similar to the existing implementation for Tasks and Pipelines. The when field will accept a list of When Expressions that define the conditions for executing the Step.
+
+### When Expressions Evaluation
+
+When expressions are evaluated at the entrypointer level. This allows us to use step results produced from previous steps to determine if a step should be skipped. To indicate a skipped step, the exit code will be 0 (showing a step is complete) and the TerminationReason will be Skipped.
+
+### Supported Expression Languages
+
+Just like when expressions in Task Level, when expressions in Step Level will also be supporting both Operator-based expressions and CEL(Common Expression Language) expressions. 
```

Comment:
May be also add a paragraph on which fields are we applying when expressions on?
- all task based fields(e.g. Task Results, Task Conditions etc.)
- StepResults
- StepStatus(?): `TerminationReasonFields: Error, Continued, Skipped, TimeOut, Cancelled etc.`

Reply:
I added a `Variable Substitution` section to explains what which fields can be referenced in `step.when`.
 > StepStatus(?)
 It seems that we haven't StepStatus implemented yet, should we include in this tep? This seems a different scope.

Reply:
I'm ok keeping it out for now.

Reply:
IMHO, we should include some way to allow conditional execution on step status. Otherwise usage of When expression in guarding a `step` will be limited. I think one major advantage of a When expression guarding a step is to check the status of preview steps. Otherwise we are forcing the user to "mimic" this functionality by exporting a pseudo status of step to workspace/result to mimic this functionality. 

Secondly, is it possible to check attributes of files in a workspace in When expressions?

Reply:
Hi, @haroonc 
Thanks for the review! I've added step status variable substitution when evaluating When expressions in the design. 
For checking file attributes, as far as i know Tekton doesn't support that directly. But, you can use a script in a step to get attributes and store them as results. Then, use those results in When expressions in later steps.